src/nsolid/nsolid_heap_snapshot.cc

```diff
@@ -350,5 +402,169 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
   }
 }
 
+void NSolidHeapSnapshot::start_sampling_profiler(
+    SharedEnvInst envinst,
+    uint64_t duration,
+    NSolidHeapSnapshot* snapshotter) {
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_sampling_);
+  auto it = snapshotter->threads_running_heap_sampling_.find(thread_id);
+  ASSERT(it != snapshotter->threads_running_heap_sampling_.end());
+
+  HeapSamplerStor& stor = it->second;
+
+  v8::Isolate* isolate = envinst->isolate();
+  v8::HeapProfiler* profiler = isolate->GetHeapProfiler();
+  ASSERT_NOT_NULL(profiler);
+
+  profiler->StartSamplingHeapProfiler(
+      stor.sample_interval,
+      stor.stack_depth,
+      stor.sampling_flags);
+
+  // Schedule a timer to take the snapshot
+  int er = QueueCallback(duration,
+                         stop_sampling_profiler,
+                         envinst,
+                         snapshotter);
+
+  if (er) {
+    // In case the the thread is already gone, the cpu profile will be stopped
+    // on RemoveEnv, so do nothing here.
+  }
+}
+
+void NSolidHeapSnapshot::stop_sampling_profiler(
+    SharedEnvInst envinst,
+    NSolidHeapSnapshot* snapshotter) {
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_sampling_);
+  auto it = snapshotter->threads_running_heap_sampling_.find(thread_id);
+  if (it == snapshotter->threads_running_heap_sampling_.end()) {
+    // The profiler was stopped before the timer was triggered
+    return;
+  }
+
```

Comment:
We can hit the same issue here that we had with heap profiles: by the time this function is called the original sampling operation might be over and a new one might be in progress. We need unique ids.

Reply:
True that, fixed!